### ISSUE TITLE
[no-redeclare] stance conflicts with [no-shadow] (when a type has the same name as a variable)

### ISSUE THREAD

**Issue by acshef** [typescript-eslint/typescript-eslint#2790]
The [docs for `no-redeclare`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-redeclare.md) are very opinionated, stating that:
> this rule will report if you name a type and a variable the same name. **_This is intentional._** Declaring a variable and a type and a variable the same is usually an accident, and it can lead to hard-to-understand code.

Yet the [docs for `no-shadow`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-shadow.md) state:
> ### `ignoreTypeValueShadow`
> When set to `true`, the rule will ignore the case when you name a type the same as a variable.
> 
> TypeScript allows types and variables to shadow one-another. This is generally safe because you cannot use variables in type locations without a `typeof` operator, so there's little risk of confusion.

**I expect the docs to have a consistent stance. Furthermore, I personally expect that stance to be such that the rules allow me to name a variable and a type the same thing, while still gaining the others benefits that `no-redeclare` and `no-shadow` offer. I'd like the option to truly ignore types and variables with the same name in `no-redeclare`.**

---

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-redeclare": [
      "error",
      {
        "builtinGlobals": true,
        "ignoreDeclarationMerge": true
      }
    ]
  },
  "@typescript-eslint/no-shadow": [
    "error",
    {
      "ignoreTypeValueShadow": true
    }
  ]
}
```

```TS
export namespace Foo {
  export interface Bar {
    // ...
  }
}

export const Foo: any = true;
```

**Expected Result**

No error.

**Actual Result**

```
1:14  error  'Foo' is already defined    @typescript-eslint/no-redeclare
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.8.1` |
| `@typescript-eslint/parser`        | `4.8.1` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.13.0` |
| `node`                             | `14.6.0` |

**Comment by bradzacher** [comment]
Shadowing and redeclaring are similar but different things.

```
const x = 1;
const x = 2; // x has been redeclared.

const y = 3;
function foo() {
  const y = 4; // y has been shadowed.
}
```

For many people, shadowing is considered okay. For an engine, shadowing is 100% okay.
This is why shadowing is much more lenient and widely accepted by people.
In addition to that, before I rewrote the scope analysis the rule used to allow it. This is the primary reason the option exists - so people don't get a bunch of errors.

Redeclaration OTOH is a lot less acceptable. Whilst technically, yes, it is valid in TS to name a type and a value the same, the vast, vast majority of the time it is incorrect code. This rule had no legacy to continue, hence an option was not added.

**Comment by acshef** [comment]
@bradzacher then perhaps the docs for `no-shadow` are wrong/misguided?
> Examples of correct code with `{ ignoreTypeValueShadow: true }`:
> 
> ```TS
> type Foo = number;
> const Foo = 1;
>
> interface Bar {
>   prop: number;
> }
> const Bar = 'test';
> ```

This looks like declaration merging, not shadowing.
